In [15]:
import pandas as pd
import numpy as np

In [16]:
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
orders.shape

(99441, 8)

In [24]:
print(orders["order_id"].nunique())

print(orders["customer_id"].nunique())

print(orders["order_status"].value_counts())

99441
99441
order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64


## Orders Table — Initial Understanding

- Number of rows: 99,441
- Number of columns: 8
- Grain: One row represents one order
- Primary key: order_id
- customer_id identifies the customer who placed the order
- order_status represents the order's status
- The table contains multiple timestamps related to the order lifecycle

In [26]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
customers.shape

(99441, 5)

In [31]:
print(customers.head())
print(customers.shape)
print(customers.columns)
print(customers.info())

                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
(99441, 5)
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')
<class '

## Customers Table — Initial Understanding

- Number of rows: 99,441
- Number of columns: 5
- `customer_id` identifies a customer record
- `customer_unique_id` identifies the underlying unique customer
- Customer location is represented by ZIP-code prefix, city, and state
- No missing values were observed during the initial inspection
- `customer_id` and `customer_unique_id` must be distinguished when performing customer-level analysis

In [32]:
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")

In [34]:
print(order_items.head())
print(order_items.shape)
print(order_items.columns)
print(order_items.info())

                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.7

In [37]:
print(order_items["order_id"].nunique())
print(order_items["order_id"].duplicated().sum())

98666
13984


## Order Items Table — Initial Understanding

- Number of rows: 112,650
- Number of columns: 7
- Grain: One row represents one product line within an order
- `order_id` is not unique because an order can contain multiple items
- `(order_id, order_item_id)` identifies an individual item within an order
- `product_id` identifies the purchased product
- `seller_id` identifies the seller
- `price` represents the item price
- `freight_value` represents the freight/shipping value
- `shipping_limit_date` represents the seller's shipping deadline
- No missing values were observed during the initial inspection

In [38]:
products = pd.read_csv("../data/raw/olist_products_dataset.csv")

In [44]:
print(products.head())
print(products.shape)
print(products.columns)
print(products.info())

                         product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0       

## Products Table — Initial Understanding

- Number of rows: 32,951
- Number of columns: 9
- Grain: One row represents one product
- `product_id` identifies the product
- `product_category_name` identifies the product category
- Product attributes include name length, description length, number of photos, weight, and dimensions
- Missing values were observed in product category and descriptive attributes
- A small number of products also have missing physical dimensions/weight
- Product categories are stored in Portuguese and can be translated using the category translation table

In [45]:
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")

In [46]:
print(sellers.head())
print(sellers.shape)
print(sellers.columns)
print(sellers.info())

                          seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP  
(3095, 4)
Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='str')
<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   

## Sellers Table — Initial Understanding

- Number of rows: 3,095
- Number of columns: 4
- Grain: One row represents one seller
- `seller_id` identifies the seller
- Seller location is represented by ZIP-code prefix, city, and state
- No missing values were observed during the initial inspection
- `seller_id` connects sellers to the order items table

In [48]:
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")

In [49]:
print(payments.head())
print(payments.shape)
print(payments.columns)
print(payments.info())

                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  
(103886, 5)
Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')
<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                ------

## Payments Table — Initial Understanding

- Number of rows: 103,886
- Number of columns: 5
- Payment records are linked to orders using `order_id`
- An order can have multiple payment records
- `payment_sequential` represents the payment sequence within an order
- `payment_type` identifies the payment method
- `payment_installments` represents the number of installments
- `payment_value` represents the payment amount
- No missing values were observed during the initial inspection
- The number of payment records is greater than the number of orders, indicating that some orders have multiple payment records

In [50]:
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")

In [52]:
print(reviews.head())
print(reviews.shape)
print(reviews.columns)
print(reviews.info())

                          review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3           

## Reviews Table — Initial Understanding

- Number of rows: 99,224
- Number of columns: 7
- Grain: One row represents a customer review record
- `review_id` identifies the review
- `order_id` connects the review to an order
- `review_score` represents the customer's rating
- Review comments are optional and contain substantial missing values
- Missing review comments should not automatically be treated as invalid records
- `review_creation_date` and `review_answer_timestamp` are currently stored as strings
- The table can be used to analyze customer satisfaction and its relationship with delivery and sales factors

In [53]:
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")

In [54]:
print(geolocation.head())
print(geolocation.shape)
print(geolocation.columns)
print(geolocation.info())

   geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  
(1000163, 5)
Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='str')
<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       -----

## Geolocation Table — Initial Understanding

- Number of rows: 1,000,163
- Number of columns: 5
- The table contains geographic information associated with ZIP-code prefixes
- Geographic attributes include latitude, longitude, city, and state
- No missing values were observed during the initial inspection
- A ZIP-code prefix can appear in multiple rows, so rows should not automatically be treated as unique locations
- The table can support geographic and regional analysis
- Its large size means it should only be used when it provides meaningful analytical value

In [56]:
category_translation = pd.read_csv(
    "../data/raw/product_category_name_translation.csv"
)

In [57]:
print(category_translation.head())
print(category_translation.shape)
print(category_translation.columns)
print(category_translation.info())

    product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor
(71, 2)
Index(['product_category_name', 'product_category_name_english'], dtype='str')
<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB
None


## Category Translation Table — Initial Understanding

- Number of rows: 71
- Number of columns: 2
- The table maps Portuguese product category names to English names
- No missing values were observed during the initial inspection
- `product_category_name` can be used to connect this table with the Products table
- The English category names will be useful for business reporting and dashboard visualization

# Data Quality Assessment

In [59]:
# Check missing values in each major table

tables = {
    "orders": orders,
    "customers": customers,
    "order_items": order_items,
    "products": products,
    "sellers": sellers,
    "payments": payments,
    "reviews": reviews,
    "geolocation": geolocation,
    "category_translation": category_translation
}

## Missing Value Assessment

In [60]:
for name, df in tables.items():
    print(f"\n{name.upper()}")
    print(df.isnull().sum())


ORDERS
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

CUSTOMERS
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

ORDER_ITEMS
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

PRODUCTS
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm  

## Duplicate Record Assessment

In [61]:
for name, df in tables.items():
    print(f"{name.upper()}: {df.duplicated().sum()} duplicate rows")

ORDERS: 0 duplicate rows
CUSTOMERS: 0 duplicate rows
ORDER_ITEMS: 0 duplicate rows
PRODUCTS: 0 duplicate rows
SELLERS: 0 duplicate rows
PAYMENTS: 0 duplicate rows
REVIEWS: 0 duplicate rows
GEOLOCATION: 261831 duplicate rows
CATEGORY_TRANSLATION: 0 duplicate rows


In [64]:
geolocation["geolocation_zip_code_prefix"].nunique()
geolocation["geolocation_zip_code_prefix"].value_counts().head(10)

geolocation_zip_code_prefix
24220    1146
24230    1102
38400     965
35500     907
11680     879
22631     832
30140     810
11740     788
38408     773
28970     743
Name: count, dtype: int64

### Geolocation Duplicate Finding

- The Geolocation table contains 261,831 exact duplicate rows.
- There are 19,015 unique ZIP-code prefixes among 1,000,163 records.
- Some ZIP-code prefixes occur hundreds of times.
- Repeated geographic records should not be blindly deleted because the table may contain multiple geographic observations for the same ZIP-code prefix.
- Geolocation will be treated as an optional supporting dataset rather than a core dependency for the MVP.

In [65]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Order Status Assessment

- All observed order statuses appear to represent valid stages or outcomes in the order lifecycle.
- `delivered` is the dominant order status.
- Orders with statuses such as `shipped`, `processing`, `created`, and `approved` represent orders that were not yet delivered.
- `canceled` and `unavailable` represent unsuccessful order outcomes.
- Order status should be considered when defining completed orders and business KPIs.
- No unexpected order status values were identified during the initial assessment.

In [68]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

### Review Score Assessment

- `review_score` contains values only from 1 to 5.
- No invalid or unexpected review scores were identified.
- No missing review scores were observed.
- The distribution is heavily concentrated around score 5.
- The score distribution will be analyzed later as part of customer satisfaction analysis.

In [69]:
payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

### Payment Type Assessment

- The dataset contains four main payment methods: credit card, boleto, voucher, and debit card.
- `not_defined` appears in only 3 records.
- The `not_defined` value may represent an undefined payment method rather than an invalid record.
- These records should be investigated before deciding whether to keep, transform, or exclude them.
- No widespread unexpected payment method values were identified.

In [70]:
payments[payments["payment_type"] == "not_defined"]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0


### Not-Defined Payment Records

- Three payment records have `payment_type = not_defined`.
- All three records have `payment_value = 0.0`.
- These records appear to represent undefined or zero-value payment transactions.
- Because they represent only a very small number of records, they are unlikely to materially affect overall payment analysis.
- They will be reviewed during the cleaning stage rather than being removed during the initial quality assessment.

In [71]:
payments["payment_installments"].value_counts().sort_index()

payment_installments
0         2
1     52546
2     12413
3     10461
4      7098
5      5239
6      3920
7      1626
8      4268
9       644
10     5328
11       23
12      133
13       16
14       15
15       74
16        5
17        8
18       27
20       17
21        3
22        1
23        1
24       18
Name: count, dtype: int64

### Payment Installments Assessment

- Most payment records contain between 1 and 24 installments.
- Two records have `payment_installments = 0`.
- Zero installments may be associated with undefined or zero-value payment records.
- No negative installment values were identified.
- The two zero-installment records will be investigated before cleaning.
- Other installment values appear reasonable for payment analysis.

In [73]:
payments[payments["payment_installments"] == 0]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


### Zero-Installment Payment Records

- Two credit-card payment records have `payment_installments = 0`.
- Both records have a positive `payment_value`.
- These records differ from the `not_defined` payment records because they represent valid credit-card payments with non-zero values.
- `payment_installments = 0` is therefore treated as a potential data-quality anomaly.
- The records will be reviewed during the cleaning stage rather than being removed during assessment.

In [74]:
payments["payment_value"].describe()

count    103886.000000
mean        154.100380
std         217.494064
min           0.000000
25%          56.790000
50%         100.000000
75%         171.837500
max       13664.080000
Name: payment_value, dtype: float64

No negative payment records is found

In [76]:
order_items[["price", "freight_value"]].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


### Order Item Price and Freight Assessment

- No missing values were identified for `price` or `freight_value`.
- All product prices are positive, with a minimum of 0.85.
- `freight_value` contains some zero values, which may represent free shipping.
- The maximum price is 6,735.00 and the maximum freight value is 409.68.
- High values are not automatically considered invalid because they may represent legitimate products or shipping costs.
- No obvious impossible values were identified during the initial assessment.

In [77]:
products[
    [
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
].describe()

,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32949.000000,32949.000000,32949.000000,32949.000000
mean,2276.472488,30.815078,16.937661,23.196728
std,4282.038731,16.914458,13.637554,12.079047
min,0.000000,7.000000,2.000000,6.000000
25%,300.000000,18.000000,8.000000,15.000000
50%,700.000000,25.000000,13.000000,20.000000
75%,1900.000000,38.000000,21.000000,30.000000
max,40425.000000,105.000000,105.000000,118.000000


### Product Physical Attributes Assessment

- Product physical measurements contain 2 missing records.
- No negative values were identified for weight or dimensions.
- Product dimensions are all positive.
- `product_weight_g` contains a minimum value of 0.
- Zero product weight is unusual and should be investigated before cleaning.
- The maximum weight is 40,425 g, which may represent a legitimate heavy product.
- No obviously impossible dimension values were identified during the initial assessment.

In [78]:
products[products["product_weight_g"] == 0]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
9769,81781c0fed9fe1ad6e8c81fca1e1cb08,cama_mesa_banho,51.0,529.0,1.0,0.0,30.0,25.0,30.0
13683,8038040ee2a71048d4bdbbdc985b69ab,cama_mesa_banho,48.0,528.0,1.0,0.0,30.0,25.0,30.0
14997,36ba42dd187055e1fbe943b2d11430ca,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0
32079,e673e90efa65a5409ff4196c038bb5af,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0


### Zero-Weight Product Records

- Four products have `product_weight_g = 0`.
- All four products have valid positive dimensions.
- All four belong to the `cama_mesa_banho` product category.
- Because the products have physical dimensions but zero recorded weight, the zero-weight values may represent missing or incorrect measurements.
- These records will be reviewed during the cleaning stage rather than being automatically deleted or assigned an estimated weight.

In [79]:
products["product_photos_qty"].value_counts().sort_index()

product_photos_qty
1.0     16489
2.0      6263
3.0      3860
4.0      2428
5.0      1484
6.0       968
7.0       343
8.0       192
9.0       105
10.0       95
11.0       46
12.0       35
13.0        9
14.0        5
15.0        8
17.0        7
18.0        2
19.0        1
20.0        1
Name: count, dtype: int64

### Product Photos Assessment

- `product_photos_qty` contains values from 1 to 20.
- No zero or negative photo counts were identified.
- No unexpected fractional values were identified.
- 610 product records have missing photo counts.
- Missing photo counts should not automatically be converted to zero because the absence of a recorded value does not necessarily mean the product had no photos.
- The missing values will be handled during the cleaning stage.

In [80]:
products[
    ["product_name_lenght", "product_description_lenght"]
].describe()

,product_name_lenght,product_description_lenght
count,32341.000000,32341.000000
mean,48.476949,771.495285
std,10.245741,635.115225
min,5.000000,4.000000
25%,42.000000,339.000000
50%,51.000000,595.000000
75%,57.000000,972.000000
max,76.000000,3992.000000


In [83]:
print("Orders:", orders["order_id"].nunique(), len(orders))
print("Customers:", customers["customer_id"].nunique(), len(customers))
print("Products:", products["product_id"].nunique(), len(products))
print("Sellers:", sellers["seller_id"].nunique(), len(sellers))
print("Reviews:", reviews["review_id"].nunique(), len(reviews))
print("Category Translation:", category_translation["product_category_name"].nunique(), len(category_translation))

Orders: 99441 99441
Customers: 99441 99441
Products: 32951 32951
Sellers: 3095 3095
Reviews: 98410 99224
Category Translation: 71 71


### Primary Key Assessment

- `order_id`, `customer_id`, `product_id`, and `seller_id` are unique within their respective tables.
- `product_category_name` is unique within the Category Translation table.
- `review_id` is not unique: 98,410 unique review IDs exist across 99,224 rows.
- The Reviews table therefore requires further investigation before `review_id` can be treated as a primary key.
- No records were removed based on this finding during the quality assessment.

In [84]:
reviews[reviews.duplicated("review_id", keep=False)].sort_values("review_id").head(20)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
57280,0174caf0ee5964646040cd94e15ac95e,74db91e33b4e1fd865356c89a61abf1f,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
54832,017808d29fd1f942d97e50184dfb4c13,8daaa9e99d60fbba579cc1c3e3bfae01,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
99167,017808d29fd1f942d97e50184dfb4c13,b1461c8882153b5fe68307c46a506e39,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
20621,0254bd905dc677a6078990aad3331a36,5bf226cf882c5bf4247f89a97c86f273,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44
96080,0254bd905dc677a6078990aad3331a36,331b367bdd766f3d1cf518777317b5d9,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44


In [85]:
review_id_check = (
    reviews.groupby("review_id")["order_id"]
    .nunique()
    .value_counts()
    .sort_index()
)

review_id_check

order_id
1    97621
2      764
3       25
Name: count, dtype: int64

In [86]:
reviews["order_id"].nunique(), len(reviews)

(98673, 99224)

### Reviews Order Relationship Assessment

- The Reviews table contains 99,224 records and 98,673 unique `order_id` values.
- Therefore, some orders have multiple review records.
- `order_id` is not unique within the Reviews table.
- Combined with the earlier finding that `review_id` is also not unique, the Reviews table does not have a simple single-column primary key.
- Reviews will be linked to the Orders table using `order_id`.
- Multiple review records for an order will not be removed automatically and will be handled carefully during analysis.

In [87]:
order_items[["order_id", "order_item_id"]].duplicated().sum()

np.int64(0)

### Order Items Key Assessment

- `order_id` is not unique because an order can contain multiple products.
- `order_item_id` identifies the item sequence within an order.
- The combination of `order_id` and `order_item_id` contains 0 duplicate records.
- Therefore, (`order_id`, `order_item_id`) can be treated as the composite key for the Order Items table.

In [88]:
id_check = orders[~orders["customer_id"].isin(customers["customer_id"])]

print("Missing customer references:", len(id_check))

Missing customer references: 0


### Orders → Customers Foreign Key Assessment

- All `customer_id` values in the Orders table have a matching record in the Customers table.
- No missing customer references were identified.
- The `Orders.customer_id → Customers.customer_id` relationship is valid.

In [89]:
missing_products = order_items[
    ~order_items["product_id"].isin(products["product_id"])
]

print("Missing product references:", len(missing_products))

Missing product references: 0


### Order Items → Products Foreign Key Assessment

- All `product_id` values in Order Items have a matching record in the Products table.
- No missing product references were identified.
- The `Order Items.product_id → Products.product_id` relationship is valid.

In [90]:
missing_sellers = order_items[
    ~order_items["seller_id"].isin(sellers["seller_id"])
]

print("Missing seller references:", len(missing_sellers))

Missing seller references: 0


### Order Items → Sellers Foreign Key Assessment

- All `seller_id` values in Order Items have a matching record in the Sellers table.
- No missing seller references were identified.
- The `Order Items.seller_id → Sellers.seller_id` relationship is valid.

In [91]:
missing_orders_from_payments = payments[
    ~payments["order_id"].isin(orders["order_id"])
]

print("Missing order references:", len(missing_orders_from_payments))

Missing order references: 0


### Payments → Orders Foreign Key Assessment

- All `order_id` values in Payments have a matching record in the Orders table.
- No missing order references were identified.
- The `Payments.order_id → Orders.order_id` relationship is valid.

In [92]:
missing_orders_from_reviews = reviews[
    ~reviews["order_id"].isin(orders["order_id"])
]

print("Missing order references:", len(missing_orders_from_reviews))

Missing order references: 0


### Reviews → Orders Foreign Key Assessment

- All `order_id` values in Reviews have a matching record in the Orders table.
- No missing order references were identified.
- The `Reviews.order_id → Orders.order_id` relationship is valid.
- Multiple review records may exist for the same order and will be handled during analysis.

In [93]:
missing_categories = products[
    products["product_category_name"].notna()
    & ~products["product_category_name"].isin(
        category_translation["product_category_name"]
    )
]

print("Missing category translations:", len(missing_categories))

Missing category translations: 13


In [94]:
missing_categories["product_category_name"].value_counts()

product_category_name
portateis_cozinha_e_preparadores_de_alimentos    10
pc_gamer                                          3
Name: count, dtype: int64

### Products → Category Translation Assessment

- 13 product records have categories that are not present in the Category Translation table.
- The missing translations belong to two categories:
  - `portateis_cozinha_e_preparadores_de_alimentos` — 10 products
  - `pc_gamer` — 3 products
- These records are not considered invalid products; only their English translations are missing.
- The original category values will be preserved.
- During cleaning, available categories will be translated while these two categories will be handled separately rather than being deleted.

In [95]:
missing_orders_from_items = order_items[
    ~order_items["order_id"].isin(orders["order_id"])
]

print("Missing order references:", len(missing_orders_from_items))

Missing order references: 0


In [96]:
orders["order_purchase_timestamp"] = pd.to_datetime(
    orders["order_purchase_timestamp"]
)

orders["order_delivered_customer_date"] = pd.to_datetime(
    orders["order_delivered_customer_date"]
)

invalid_delivery_dates = orders[
    orders["order_delivered_customer_date"].notna()
    & (
        orders["order_delivered_customer_date"]
        < orders["order_purchase_timestamp"]
    )
]

print("Delivered before purchase:", len(invalid_delivery_dates))

Delivered before purchase: 0


In [97]:
orders["order_approved_at"] = pd.to_datetime(
    orders["order_approved_at"]
)

invalid_approval_dates = orders[
    orders["order_approved_at"].notna()
    & (
        orders["order_approved_at"]
        < orders["order_purchase_timestamp"]
    )
]

print("Approved before purchase:", len(invalid_approval_dates))

Approved before purchase: 0


### Order Approval Date Validation

- No orders were found where payment approval occurred before the order was purchased.
- `order_approved_at` is chronologically valid relative to `order_purchase_timestamp`.
- No records require correction based on this check.

In [98]:
invalid_delivery_sequence = orders[
    orders["order_delivered_customer_date"].notna()
    & orders["order_delivered_carrier_date"].notna()
    & (
        orders["order_delivered_customer_date"]
        < orders["order_delivered_carrier_date"]
    )
]

print("Customer delivery before carrier handover:", len(invalid_delivery_sequence))

Customer delivery before carrier handover: 23


In [99]:
invalid_delivery_sequence[
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
6437,a1abeb653a4d4cd1e142ccb8c82cd069,delivered,2017-07-20 11:20:52,2017-07-21 06:43:14,2017-07-28 16:57:58,2017-07-25 19:32:56,2017-08-14 00:00:00
9553,383aa8b2724fe452d9ccd9934a8c628b,delivered,2017-07-02 20:58:43,2017-07-02 21:10:20,2017-07-07 17:22:41,2017-07-06 14:27:51,2017-07-21 00:00:00
13487,cb1134f9010d242e9515ad1c78ec0c39,delivered,2017-07-16 12:35:34,2017-07-18 06:03:50,2017-07-20 19:22:02,2017-07-19 14:13:28,2017-08-08 00:00:00
14474,dceb62e8fa94b46006c9554fed743df0,delivered,2017-07-20 20:58:05,2017-07-22 11:45:11,2017-08-01 18:23:30,2017-07-26 18:09:10,2017-08-11 00:00:00
19268,5f9d46795c3126674e52becb3a1a517f,delivered,2017-07-18 11:48:20,2017-07-18 12:03:29,2017-07-20 23:03:42,2017-07-20 18:52:41,2017-07-31 00:00:00
21338,8c78d01de3a9009e23d6877a7cc9be20,delivered,2016-10-08 15:36:50,2016-10-08 18:13:44,2016-10-26 11:41:53,2016-10-25 17:51:46,2016-11-30 00:00:00
22520,b27af682321527a6349f1761eb3f360c,delivered,2017-06-14 20:17:04,2017-06-14 20:30:08,2017-06-27 14:51:54,2017-06-26 15:45:35,2017-07-14 00:00:00
25393,1cc3ae63caffff2d6c3ee3e78e074acf,delivered,2017-08-07 21:35:22,2017-08-08 21:45:15,2017-08-10 18:28:56,2017-08-10 18:05:38,2017-08-25 00:00:00
25646,e37f11cae9985ca58f0b56f268720537,delivered,2017-07-26 11:46:34,2017-07-27 10:10:16,2017-08-01 18:17:47,2017-07-31 17:49:56,2017-08-24 00:00:00
27470,fa3e37584f4fdb1ded0e0de700dfcb4e,delivered,2017-07-30 19:32:23,2017-07-30 19:45:09,2017-08-09 18:18:43,2017-08-01 21:13:01,2017-08-18 00:00:00


### Delivery Sequence Anomaly

- 23 orders have a customer delivery timestamp earlier than the carrier handover timestamp.
- These records have otherwise plausible order lifecycle dates.
- The anomaly represents approximately 0.02% of all orders.
- The records will not be deleted because the source data should be preserved.
- These records will be flagged as delivery-sequence anomalies and excluded only from analyses where this inconsistent sequence could affect the calculation.

In [100]:
late_deliveries = orders[
    orders["order_delivered_customer_date"].notna()
    & (
        orders["order_delivered_customer_date"]
        > orders["order_estimated_delivery_date"]
    )
]

print("Orders delivered after estimated date:", len(late_deliveries))

Orders delivered after estimated date: 7827


### Estimated Delivery Date Assessment

- 7,827 delivered orders have a customer delivery date later than the estimated delivery date.
- These records are not considered data-quality errors because late delivery is a valid business outcome.
- Late delivery will be retained as a business metric for operational analysis.
- This metric can be used to evaluate delivery performance and its relationship with customer satisfaction.

In [101]:
for name, df in tables.items():
    print(f"\n{name.upper()}")
    print(df.dtypes)


ORDERS
order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date                str
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date               str
dtype: object

CUSTOMERS
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

ORDER_ITEMS
order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

PRODUCTS
product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    f

In [102]:
orders["order_delivered_carrier_date"] = pd.to_datetime(
    orders["order_delivered_carrier_date"]
)

orders["order_estimated_delivery_date"] = pd.to_datetime(
    orders["order_estimated_delivery_date"]
)

order_items["shipping_limit_date"] = pd.to_datetime(
    order_items["shipping_limit_date"]
)

reviews["review_creation_date"] = pd.to_datetime(
    reviews["review_creation_date"]
)

reviews["review_answer_timestamp"] = pd.to_datetime(
    reviews["review_answer_timestamp"]
)

print("Date columns converted successfully.")

Date columns converted successfully.


In [103]:
print("Orders:")
print(orders[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].dtypes)

print("\nOrder Items:")
print(order_items["shipping_limit_date"].dtype)

print("\nReviews:")
print(reviews[
    [
        "review_creation_date",
        "review_answer_timestamp"
    ]
].dtypes)

Orders:
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

Order Items:
datetime64[us]

Reviews:
review_creation_date       datetime64[us]
review_answer_timestamp    datetime64[us]
dtype: object


### Data Type Validation

- All order lifecycle date columns have been successfully converted to datetime.
- `shipping_limit_date` in Order Items has been converted to datetime.
- Review creation and answer timestamps have been converted to datetime.
- ID columns remain strings, which is appropriate for identifiers.
- Numeric measures remain numeric.
- No data-type issues requiring further correction were identified.

In [104]:
numeric_checks = {
    "price": order_items["price"],
    "freight_value": order_items["freight_value"],
    "payment_value": payments["payment_value"],
    "payment_installments": payments["payment_installments"],
    "review_score": reviews["review_score"],
    "product_weight_g": products["product_weight_g"],
    "product_length_cm": products["product_length_cm"],
    "product_height_cm": products["product_height_cm"],
    "product_width_cm": products["product_width_cm"]
}

for column, series in numeric_checks.items():
    print(f"{column}: { (series < 0).sum() } negative values")

price: 0 negative values
freight_value: 0 negative values
payment_value: 0 negative values
payment_installments: 0 negative values
review_score: 0 negative values
product_weight_g: 0 negative values
product_length_cm: 0 negative values
product_height_cm: 0 negative values
product_width_cm: 0 negative values


### Negative Value Assessment

- No negative values were found in price or freight value.
- No negative payment values were found.
- No negative installment values were found.
- No negative review scores were found.
- No negative product weight or dimension values were found.
- No records require correction based on negative-value validation.